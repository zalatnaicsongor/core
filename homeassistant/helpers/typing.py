"""Typing Helpers for Home Assistant."""

from collections.abc import Mapping
from enum import Enum
from functools import partial
from typing import Any, Never

import homeassistant.core

from .deprecation import (
    DeprecatedAlias,
    all_with_deprecated_constants,
    check_if_deprecated_constant,
    dir_with_deprecated_constants,
)

GPSType = tuple[float, float]
ConfigType = dict[str, Any]
DiscoveryInfoType = dict[str, Any]
ServiceDataType = dict[str, Any]
StateType = str | int | float | None
TemplateVarsType = Mapping[str, Any] | None
NoEventData = Mapping[str, Never]

# Custom type for recorder Queries
QueryType = Any


class UndefinedType(Enum):
    """Singleton type for use with not set sentinel values."""

    _singleton = 0


UNDEFINED = UndefinedType._singleton  # noqa: SLF001


# The following types should not used and
# are not present in the core code base.
# They are kept in order not to break custom integrations
# that may rely on them.
# Deprecated as of 2024.5 use types from homeassistant.core instead.
_DEPRECATED_ContextType = DeprecatedAlias(
    homeassistant.core.Context, "homeassistant.core.Context", "2025.5"
)
_DEPRECATED_EventType = DeprecatedAlias(
    homeassistant.core.Event, "homeassistant.core.Event", "2025.5"
)
_DEPRECATED_HomeAssistantType = DeprecatedAlias(
    homeassistant.core.HomeAssistant, "homeassistant.core.HomeAssistant", "2025.5"
)
_DEPRECATED_ServiceCallType = DeprecatedAlias(
    homeassistant.core.ServiceCall, "homeassistant.core.ServiceCall", "2025.5"
)

# These can be removed if no deprecated constant are in this module anymore
__getattr__ = partial(check_if_deprecated_constant, module_globals=globals())
__dir__ = partial(
    dir_with_deprecated_constants, module_globals_keys=[*globals().keys()]
)
__all__ = all_with_deprecated_constants(globals())
