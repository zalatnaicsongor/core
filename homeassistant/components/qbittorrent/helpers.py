"""Helper functions for qBittorrent."""

from datetime import UTC, datetime
from typing import Any

from qbittorrent.client import Client


def setup_client(url: str, username: str, password: str, verify_ssl: bool) -> Client:
    """Create a qBittorrent client."""
    client = Client(url, verify=verify_ssl)
    client.login(username, password)
    # Get an arbitrary attribute to test if connection succeeds
    client.get_alternative_speed_status()
    return client


def seconds_to_hhmmss(seconds) -> str:
    """Convert seconds to HH:MM:SS format."""
    if seconds == 8640000:
        return "None"

    minutes, seconds = divmod(seconds, 60)
    hours, minutes = divmod(minutes, 60)
    return f"{int(hours):02}:{int(minutes):02}:{int(seconds):02}"


def format_unix_timestamp(timestamp) -> str:
    """Format a UNIX timestamp to a human-readable date."""
    dt_object = datetime.fromtimestamp(timestamp, tz=UTC)
    return dt_object.isoformat()


def format_progress(torrent) -> str:
    """Format the progress of a torrent."""
    progress = torrent["progress"]
    progress = float(progress) * 100
    return f"{progress:.2f}"


def format_torrents(torrents: list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
    """Format a list of torrents."""
    value = {}
    for torrent in torrents:
        value[torrent["name"]] = format_torrent(torrent)

    return value


def format_torrent(torrent) -> dict[str, Any]:
    """Format a single torrent."""
    value = {}
    value["id"] = torrent["hash"]
    value["added_date"] = format_unix_timestamp(torrent["added_on"])
    value["percent_done"] = format_progress(torrent)
    value["status"] = torrent["state"]
    value["eta"] = seconds_to_hhmmss(torrent["eta"])
    value["ratio"] = "{:.2f}".format(float(torrent["ratio"]))

    return value
