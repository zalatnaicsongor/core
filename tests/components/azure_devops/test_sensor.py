"""Tests for init of Azure DevOps."""

from unittest.mock import AsyncMock

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er

from . import setup_integration

from tests.common import MockConfigEntry


@pytest.mark.usefixtures("entity_registry_enabled_by_default")
async def test_sensors(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    entity_registry: er.EntityRegistry,
    mock_config_entry: MockConfigEntry,
    mock_devops_client: AsyncMock,
) -> None:
    """Test the sensor entities."""
    assert await setup_integration(hass, mock_config_entry)

    assert (
        entry := entity_registry.async_get("sensor.testproject_test_build_latest_build")
    )

    assert entry == snapshot(name=f"{entry.entity_id}-entry")

    assert hass.states.get(entry.entity_id) == snapshot(name=f"{entry.entity_id}-state")
