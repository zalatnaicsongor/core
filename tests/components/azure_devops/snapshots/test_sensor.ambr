# serializer version: 1
# name: test_sensors[sensor.testproject_test_build_latest_build-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'sensor',
    'entity_category': None,
    'entity_id': 'sensor.testproject_test_build_latest_build',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Test Build latest build',
    'platform': 'azure_devops',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'latest_build',
    'unique_id': 'testorg_1234_9876_latest_build',
    'unit_of_measurement': None,
  })
# ---
# name: test_sensors[sensor.testproject_test_build_latest_build-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'definition_id': 9876,
      'definition_name': 'Test Build',
      'finish_time': '2021-01-01T00:00:00Z',
      'friendly_name': 'testproject Test Build latest build',
      'id': 5678,
      'queue_time': '2021-01-01T00:00:00Z',
      'reason': 'manual',
      'result': 'succeeded',
      'source_branch': 'main',
      'source_version': '123',
      'start_time': '2021-01-01T00:00:00Z',
      'status': 'completed',
      'url': None,
    }),
    'context': <ANY>,
    'entity_id': 'sensor.testproject_test_build_latest_build',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': '1',
  })
# ---
